src/UiPath.Workflow.Runtime/ParallelTrackingExtensions.cs

```diff
@@ -9,23 +9,48 @@
 /// </summary>
 public static class ParallelTrackingExtensions
 {
-    public const string BranchIdPropertyName = "__ParallelBranchId";
+    private const string BranchIdPropertyName = "__ParallelBranchId";
 
     public static ActivityInstance MarkNewParallelBranch(this ActivityInstance instance)
     {
-        var properties = new ExecutionProperties(null, instance, instance.PropertyManager);
-        var parentId = properties.Find(BranchIdPropertyName) as string;
-        properties.Add(BranchIdPropertyName, $"{parentId}.{Guid.NewGuid():N}".Trim('.'), true, false);
-
+        var parentId = instance.GetCurrentParallelBranchId();
+        instance.SetCurrentParallelBranchId(PushNewBranch(parentId));
         return instance;
     }
 
     public static string GetCurrentParallelBranchId(this ActivityInstance instance) =>
-        instance.PropertyManager?.GetPropertyAtCurrentScope(BranchIdPropertyName) as string;
-
-    public static ActivityInstance MarkNewParallelBranch(this ActivityContext context) =>
-         context.CurrentInstance?.MarkNewParallelBranch();
+        GetExecutionProperties(instance).Find(BranchIdPropertyName) as string;
 
     public static string GetCurrentParallelBranchId(this ActivityContext context) =>
          context.CurrentInstance?.GetCurrentParallelBranchId();
+
+    /// <summary>
+    /// Sets the parallelBranchId for the current activity instance
+    /// </summary>
+    /// <param name="branchId">null or empty removes the branch setting from current instance</param>
+    /// <exception cref="ArgumentException">when not a pop or a push</exception>
+    public static void SetCurrentParallelBranchId(this ActivityInstance instance, string branchId)

```

Comment:
From a naming PoV, I think the "Current" part of the method name makes sense only for the ActivityContext overload which accesses the .CurrentInstance

Not sure we can rename it anymore though.

Reply:
just the setter?